### ISSUE TITLE
Bad query performance

### ISSUE THREAD

**Issue by inzanez** [arangodb/arangodb#13148]
## My Environment

* __ArangoDB Version__:        3.7.1
* __Storage Engine__:            RocksDb
* __Deployment Mode__:       Single Server
* __Deployment Strategy__:  Kubernetes, Local Storage on NVME SSD Raid
* __Configuration__:              k3s cluster, applied sysctl settings as recommended
* __Infrastructure__:              own
* __Total RAM in your machine__:        32GB
* __Disks in use__:                NVME SSD (Intel), Raid
* __Used Package__:             Docker

## Component, Query & Data

__Affected feature__:
Slow AQL query

__AQL query (if applicable)__:
```
FOR d in @@collection FILTER d.name LIKE '255' LIMIT 1250, 25 RETURN d
```
__AQL explain and/or profile (if applicable)__:
Testing response times for queries that are similar to the ones that we are going to see in the live application. We inserted 1'000'000 documents into a collection with the `name` property just being a counter ascending. The query above does run a filter on that name property.

__Problem__:
Query without filter takes around `700ms`.

__Expected result__:
I expected the query to return in less time. Testing against SQL Server / Postgres I do get the answer to such a query in half the time (sometimes even less) without any modifications.

I know that this is not the brightest query to submit, but I'm afraid that's what we have to deal with. I hoped there was something that could be done about that. We do have enough RAM, so that wouldn't be an issue.

**Comment by jsteemann** [comment]
@inzanez : there are obvious ways to speed up this query, but they require to make modifications to either the setup or the query.

One obvious improvement for the query will be to turn the `FILTER d.name LIKE '255'` into `FILTER d.name == '255'`. The original query uses the `LIKE` operator, which can do some sort of wildcard matching and is thus more expensive than running the simple equality operator (`==`).
Please note that the query currently looks for a string value of `255`, and not for the number `255`. If the intention is to look for the number `255`, the FILTER condition will be `FILTER d.name == 255` (note that this will only work if the documents in the collection actually contain numeric values and not strings).

Another improvement for the query will be to add an index on the `name` attribute. Doing so will make the query change from a full scan (looking at all documents in the collection until the limit is satisfied) into an index lookup, which will normally run a lot faster.

I suggest you look at the query execution plan for the original query, using the explain functionality in the web UI (or in the shell), then modify the query to use the adjusted filter condition and create the index. If you then compare that's query execution plan with the original one, the difference in query complexity/runtime should be obvious.

**Comment by inzanez** [comment]
@jsteemann Yes, so the intention here was to get all documents with name that has '255' in it:
* 255
* 1255
* 2255
* etc.

That's why we used the LIKE operator for the test. We also added a persistent index (unique) to the field, but strangely that didn't make any difference. Same execution time still.

**Comment by jsteemann** [comment]
Yes, in order for the index to be used, I suggest to change the `LIKE` operator to `==` as in my previous answer. If you do this, the execution time should be much better.

**Comment by inzanez** [comment]
@jsteemann But that would only return 1 document, wouldn't it? Or do I misunderstand `==` vs `LIKE`? I want to retrieve all documents that have `255` anywhere in their name...

**Comment by jsteemann** [comment]
Using `d.name LIKE '255'` and `d.name == '255'` will be equivalent. Both will look for documents with the name attribute being _equal_ to the string `255`.

If you want to perform substring matching, the query should be `d.name LIKE '%255%'` (as in SQL). I expect this query to run even a bit slower, because it will now look for any occurrence of the (sub)string `255` in the attribute, whereas previously it only needed to compare the first 3 characters of each.

Anyway, if the goal is substring matching, `==` will obviously not work, so your best bets are `LIKE` with the `%` wildcards or the `CONTAINS` AQL function: `FILTER CONTAINS(d.name, '255')`. The latter may be a bit faster than the regex-based `LIKE`, simply because it is a less powerful variant and thus less complex (read: faster).

**Comment by inzanez** [comment]
@jsteemann my mistake, it should have been `%255%`, right you are. Well, it seems as if the additional latency I see is coming from the full_count I need to retrieve and not the query itself.

**Comment by inzanez** [comment]
@jsteemann  Is there any way of speeding up these full_count retrievals?

**Comment by jsteemann** [comment]
I don't see a good way of speeding this up except by switching from LIKE to CONTAINS  However, that will still need to look at all documents in the collection, so the potential speedup will not be huge.

**Comment by jsteemann** [comment]
Another alternative is to create a view on the collection and use arangosearch. I haven't tried this yet, but it could be substantially faster.

**Comment by goedderz** [comment]
> Another alternative is to create a view on the collection and use arangosearch. I haven't tried this yet, but it could be substantially faster.

@inzanez Have a look at [n-gram analyzers](https://www.arangodb.com/docs/stable/arangosearch-analyzers.html#n-gram), they should be the right choice for this kind of lookup.

**Comment by inzanez** [comment]
@goedderz I will try that, many thanks for the hint!

**Comment by Simran-B** [comment]
Here's an (untested) example for arangosh, which should allow you to find any 3 to 6 character long substring (or the full search term):

```js
var analyzers = require("@arangodb/analyzers");
analyzers.save("my_ngrams", "ngram", { min: 3, max: 6, preserveOriginal: true, streamType: "utf8" });

db._createView("my_view", "arangosearch", { links: { my_collection: { fields: { name: { analyzers: ["my_ngrams"]}}}}});

db._query(`FOR doc IN my_collection SEARCH ANALYZER(doc.name == "255", "my_ngrams") RETURN doc`)
```
